src/net_processing.cpp

```diff
@@ -186,6 +186,51 @@ struct QueuedBlock {
     std::unique_ptr<PartiallyDownloadedBlock> partialBlock;
 };
 
+class TimeOffsets
+{
+private:
+    // Historic maximum number of time offsets we include in our median.
+    static constexpr size_t N{199};
+
+    mutable Mutex m_mutex;
+    std::array<int64_t, N> m_offsets GUARDED_BY(m_mutex){};
+    size_t m_index GUARDED_BY(m_mutex){0};
+
+public:
+    /** Add a new time offset sample. */
+    void Add(int64_t offset) EXCLUSIVE_LOCKS_REQUIRED(!m_mutex)
+    {
+        LOCK(m_mutex);
+
+        // Stop sampling after N samples.
+        // TODO This maintains a historic bug of adjusted time and could be
+        // changed in the future.
+        if (m_index >= N) return;
+
+        m_offsets[m_index] = offset;
+        ++m_index;
+    }
+
+    /** Compute and return the median of the collected time offset samples. */
+    int64_t Median() const EXCLUSIVE_LOCKS_REQUIRED(!m_mutex)
+    {
+        LOCK(m_mutex);
+
+        // Only compute the median from 5 or more samples.
+        if (m_index < 4) return 0;
```

Comment:
dde4e1c6b0cbb49b84e75b9d0d1a92161ba5a499
if `m_index` is 4, there are 4 samples but we will still compute the median, so the comment is inaccurate.